File: LeetCode - Algorithm/Linked List/328.py
# 328. Odd Even Linked List
# https://leetcode.com/problems/odd-even-linked-list/

from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        oddCurrent: Optional[ListNode] = head
        evenHead: Optional[ListNode] = head.next

        current: Optional[ListNode] = head.next
        beforeCurrent: Optional[ListNode] = head

        currentNum = 2

        while current:
            if currentNum % 2 == 1:
                beforeCurrent.next = current.next

                oddCurrent.next = current
                current.next = evenHead
                oddCurrent = oddCurrent.next

                current = beforeCurrent.next
                currentNum += 1
            else:
                beforeCurrent = current
                current = current.next
                currentNum += 1

        return head

File: LeetCode - Algorithm/Linked List/328.java
// 328. Odd Even Linked List
// https://leetcode.com/problems/odd-even-linked-list/

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    public ListNode oddEvenList(ListNode head) {
        if (head == null)
            return head;

        ListNode current = head;

        ListNode evenNodesHead = new ListNode();
        ListNode evenNodesCurrent = evenNodesHead;

        ListNode endOfOdd = current;

        while (current != null) {
            if (current.next != null) {
                evenNodesCurrent.next = current.next;

                current.next = current.next.next;

                evenNodesCurrent = evenNodesCurrent.next;

                evenNodesCurrent.next = null;
            }

            endOfOdd = current;
            current = current.next;
        }

        endOfOdd.next = evenNodesHead.next;

        return head;
    }
}